### PR TITLE
Modernize the importer file upload javascript

### LINKED ISSUE #2574: Drupal 11.x dev Issues with file upload controls in importers
### Tripal Version

4.x

### Drupal Version

11.x-dev

### PostgreSQL Version

18

### PHP Version

8.5

### Branch

`tv4g4-issue2574-tripal-file-upload-javascript`

### General Description

The "Choose file" button is not displayed in importer interfaces, preventing users from selecting files for import.

### Steps to reproduce

Navigate to Chado FASTA File Loader at `admin/tripal/loaders/chado_fasta_loader`.

On the FASTA file upload section it looks like

Image

However, it should contain the `Choose File` button as shown below

Image

Similarly, the `Choose File` button is missing on all the importers with file upload functionality.

### PR DESCRIPTION
**Pull request by dsenalik**
# Bug Fix

## Description

The bug reported in issue #2574 is due to, of course, changes in Drupal.
I think it was mostly a dependency issue in `tripal/tripal.libraries.yml` but in the process of debugging it turned into a major overhaul of the javascript code to make it more like modern Drupal 11 javascript standards.

High priority because many importers are broken on 11.x 💥!

Clipping issue actually predates 11.x, but looks like this 
4 xfasta

~~On this branch, the clipping is fixed by wrapping the markup in `<p></p>` tags~~ in commit https://github.com/tripal/tripal/pull/2576/commits/ed27d48f18e12cfb4dade6f869c99fe6038ba62c
Fixed with css, adding margin-bottom commit https://github.com/tripal/tripal/pull/2576/commits/40d84c192254c9ec28f1c98d55c8408d6a5b2078
2574fasta

An edge case, 11.x specific, occurred on the very first time a user uploaded a file.

## Testing?

**A**. We should test on Drupal 11.x-dev
**B**. and also on 10.6 to make sure that changes are backward compatible and that works too.

This needs mostly to just test that we can upload a file in an importer, fasta importer for example.
1. Tripal -> Data Loaders -> Chado Fasta File Loader
2. The "Browse..." button should be there. Click and select a fasta file.
3. Click "Upload" button
4. File should now appear in the list. 

**C**. If you really really want to test the edge case of very first upload giving an error, this only occurs on 11.x-dev. You can't test on 4.x because there is no browse button! On this branch, you would have to go to the commit just before I fixed this! And remember it only happens once on the very first upload, so this needs to be a new docker.
1. `git checkout ed27d48`
2. Tripal -> Data Loaders -> Chado Fasta File Loader
3. Click "Browse..." button select a fasta file
4. Click "Upload File" button
5. Now go to your home page (click the drupal icon in the upper left)
firstupload

5. On this branch, this will not happen, due to commit https://github.com/tripal/tripal/pull/2576/commits/36f261ce373502b773fac17326660ba058589a46 where we now create the directory before evaluating its "real" system path.
6. `git switch -` to get back to this branch

### CODE REVIEW

**Review by vidu-wil — COMMENTED** [review]
I did manual testing for this PR on both drupal 11.x-dev and 10.6 and everything works as expected.

### Drupal 11.x-dev

- I was able to reproduce the bug on `ed27d48` ✅

Image

- The error does not appear on the current branch as expected ✅

Image

- Clipped text is also fixed in the current branch ✅

Image

### Drupal 10.6

- Everything works as expected and the changes are backward compatible ✅

Image

Image

File: tripal/js/tripal.file.js [hunk 1]
```diff
@@ -1,30 +1,30 @@
...
+        const details = id.split('-');
 
-        // The settings for this uploader are provided in a custom variable
-        // specific to the table. We can get the variable name by piecing
-        // together parts of the table ID.
-        var id                = $(this).val();
-        var details           = id.split('-');
-        var settings_var_name = 'uploader_' + details[0] + '_' + details[1] + '_' + details[2];
+        const settingsVarName =
+          `uploader_${details[0]}_${details[1]}_${details[2]}`;
```
Comment (reynoldtan) [inline tripal/js/tripal.file.js:23]:
```suggestion
        const id = element.value;
        const [usage_id, usage_type, module] = id.split('-');

        const settingsVarName = `uploader_${usage_id}_${usage_type}_${module}`;

```

I've traced the components of details in this line https://github.com/tripal/tripal/blob/4992cdf42860966eb20bf137001cb6849c27f9e4/tripal/src/Element/HTML5File.php#L50. Consider this change to make this block clearer and easier to read.

File: tripal/js/tripal.file.js [hunk 2]
```diff
@@ -1,30 +1,30 @@
...
 
-        // Initialize the table for uploads.
-        tripal_files.addUploadTable(details[0] + '-' + details[1], settings.tripal[settings_var_name]);
+        if (settings.tripal?.[settingsVarName]) {
+          tripalFiles.addUploadTable(
+            `${details[0]}-${details[1]}`,
+            settings.tripal[settingsVarName]
+          );
```
Comment (reynoldtan) [inline tripal/js/tripal.file.js:28]:
```suggestion
        if (settings.tripal?.[settingsVarName]) {
          tripalFiles.addUploadTable(
            `${usage_id}-${usage_type}`,
            settings.tripal[settingsVarName]
```

Update to match my prior suggestion

File: tripal/js/tripal.file.js [hunk 3]
```diff
@@ -1,30 +1,30 @@
...
+
+      once(
+        'tripal-file-init',
+        '.tripal-html5-file-upload-table-key',
+        context
+      ).forEach((element) => {
 
-        // Set the field status
-        $(this).data('tripal.file', true);
+        const id = element.value;
```
Comment (reynoldtan) [inline tripal/js/tripal.file.js:14]:
```suggestion
        const id = element.value;
        
        if (!id.trim()) {
          return;
        }
```

Consider this check to skip ids that don't contain any value.

File: tripal/js/tripal.file.js (hunk 1 above)
Comment (dsenalik) [reply tripal/js/tripal.file.js:23]: Thanks, done! (Couldn't apply through the UI)

File: tripal/js/tripal.file.js [hunk 4]
```diff
@@ -1,30 +1,37 @@
...
-        tripal_files.addUploadTable(details[0] + '-' + details[1], settings.tripal[settings_var_name]);
+        if (settings.tripal?.[settingsVarName]) {
+          tripalFiles.addUploadTable(
+            `${details[0]}-${details[1]}`,
+            settings.tripal[settingsVarName]
+        if (settings.tripal?.[settingsVarName]) {
+          tripalFiles.addUploadTable(
+            `${usage_id}-${usage_type}`,
+            settings.tripal[settingsVarName]
+        }
```
Comment (reynoldtan) [inline tripal/js/tripal.file.js]:
```suggestion
        if (settings.tripal?.[settingsVarName]) {
          tripalFiles.addUploadTable(
            `${usage_id}-${usage_type}`,
            settings.tripal[settingsVarName]
          );
        }
```

File: tripal/js/tripal.file.js [hunk 5]
```diff
@@ -1,30 +1,37 @@
...
-        // Initialize the table for uploads.
-        tripal_files.addUploadTable(details[0] + '-' + details[1], settings.tripal[settings_var_name]);
+        if (settings.tripal?.[settingsVarName]) {
+          tripalFiles.addUploadTable(
+            `${details[0]}-${details[1]}`,
+            settings.tripal[settingsVarName]
+        if (settings.tripal?.[settingsVarName]) {
+          tripalFiles.addUploadTable(
+            `${usage_id}-${usage_type}`,
+            settings.tripal[settingsVarName]
```
Comment (reynoldtan) [inline tripal/js/tripal.file.js:27]:
```suggestion
        const uploaderSettings = settings.tripal?.[settingsVarName];

        if (uploaderSettings) {
          tripalFiles.addUploadTable(
            `${usage_id}-${usage_type}`,
            uploaderSettings
```
Consider saving the settings into a descriptive variable name for reuse and make the condition self-documenting.

**Comment by dsenalik** [comment]
All your changes looked great @reynoldtan but this commit broke the button
```
git diff 1ee0df8cb923a8..1fe9acd9c7f
diff --git a/tripal/js/tripal.file.js b/tripal/js/tripal.file.js
index 878f67be8..33c2cf328 100755
--- a/tripal/js/tripal.file.js
+++ b/tripal/js/tripal.file.js
@@ -21,7 +21,10 @@
           tripalFiles.addUploadTable(
             `${details[0]}-${details[1]}`,
             settings.tripal[settingsVarName]
-          );
+        if (settings.tripal?.[settingsVarName]) {
+          tripalFiles.addUploadTable(
+            `${usage_id}-${usage_type}`,
+            settings.tripal[settingsVarName]
         }
       });
     }
```
but I think it is probably my fault trying to get your overlapping suggestions merged.

File: tripal/js/TripalUploader.js [hunk 6]
```diff
@@ -78,236 +77,235 @@
...
-          func_name = 'live';
+          }
         }
       }
-      $(submit_id)[func_name]('click', function() {
-        for(var i = 0; i < categories.length; i++) {
-          self.start(categories[i]);
+      jQuery(submit_id)
+        .off('click.tripalUploader')
+        .on('click', () => {
```
Comment (reynoldtan) [inline tripal/js/TripalUploader.js]:
```suggestion
        .on('click.tripalUploader', () => {
```
Shouldn't this be using the same event namespace? the off() part won't be turning off anything since click != click.tripalUploader

File: tripal/js/TripalUploadFile.js [hunk 7]
```diff
@@ -178,277 +201,297 @@
...
      */
-    this.cancel = function() {
+    cancel() {
       this.status = 'cancelled';
       this.updateStatus();
     }
+
     /**
-     * 
+     *
```
Comment (reynoldtan) [inline tripal/js/TripalUploadFile.js]:
```suggestion
     * Place an upload in pending state.
```

File: tripal/src/Form/TripalImporterForm.php [hunk 8]
```diff
@@ -62,7 +62,7 @@ public function buildForm(array $form, FormStateInterface $form_state, $plugin_i
       ];
 
       $form['file']['upload_description'] = [
-        '#markup' => $importer->describeUploadFileFormat(),
+        '#markup' => '<p>' . $importer->describeUploadFileFormat() . '</p>',
```
Comment (reynoldtan) [inline tripal/src/Form/TripalImporterForm.php]: This section can contain arbitrary content such as another paragraph `<p>`, order list `<ol>`, unordered list `<ul>` or combination like in our importers [germplasm importer](https://private-user-images.githubusercontent.com/7927760/637855287-05cdc2ba-2d62-4c41-be6a-5a57a59bc0b1.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODg0NTIzNTgsIm5iZiI6MTc4ODQ1MjA1OCwicGF0aCI6Ii83OTI3NzYwLzYzNzg1NTI4Ny0wNWNkYzJiYS0yZDYyLTRjNDEtYmU2YS01YTU3YTU5YmMwYjEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDkwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjA5MDNUMTYxNDE4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2ViZDllMGZjZWI5OGZkNGNhOTk4YTU0ODQ0ZGU2MmM0NGFhYTc1NDg0ZmNkMTA3ZTBiNzJjZWI4YzFlMDY4NiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmcmVzcG9uc2UtY29udGVudC10eXBlPWltYWdlJTJGcG5nIn0.3v7zhcRdEf1Qd4BmoOUMBGnJzHXmMgt4Hpx0GXV4D0g). Consider using a neutral container such as `<div>`, and to provide a convenient hook for future styling/formatting, assign a class or id. For example

```
<div class="tripal-importer-describe-upload"> .... </div><br />
```

If clipping persists with this element, try adding `<br />` in the closing tag

**Comment by reynoldtan** [comment]
> All your changes looked great @reynoldtan but this commit broke the button
> 
> ```
> git diff 1ee0df8cb923a8..1fe9acd9c7f
> diff --git a/tripal/js/tripal.file.js b/tripal/js/tripal.file.js
> index 878f67be8..33c2cf328 100755
> --- a/tripal/js/tripal.file.js
> +++ b/tripal/js/tripal.file.js
> @@ -21,7 +21,10 @@
>            tripalFiles.addUploadTable(
>              `${details[0]}-${details[1]}`,
>              settings.tripal[settingsVarName]
> -          );
> +        if (settings.tripal?.[settingsVarName]) {
> +          tripalFiles.addUploadTable(
> +            `${usage_id}-${usage_type}`,
> +            settings.tripal[settingsVarName]
>          }
>        });
>      }
> ```
> 
> but I think it is probably my fault trying to get your overlapping suggestions merged.

All good 👍  I caught it while actively reviewing so I made a separate code suggestion for the fix

File: tripal/js/TripalUploader.js (hunk 6 above)
Comment (dsenalik) [reply tripal/js/TripalUploader.js]: Good catch - fixed and changed loop style to modernize
https://github.com/tripal/tripal/pull/2576/commits/38e3c956a065ac3e0024cb2633d18dbb4132bade

File: tripal/src/Form/TripalImporterForm.php (hunk 8 above)
Comment (dsenalik) [reply tripal/src/Form/TripalImporterForm.php]: Thanks for pointing this out. We already have an importer css library, so I used that to apply a bottom margin to the div - commit https://github.com/tripal/tripal/pull/2576/commits/40d84c192254c9ec28f1c98d55c8408d6a5b2078

**Comment by reynoldtan** [comment]
I have completed the manual test and confirmed the update fixed the issues in the importer interface.

- [x] In 11.x
image
image

- [x] in 10.6
image
image


Thanks @dsenalik for taking the time to consider my suggestions.

**Review by laceysanderson — APPROVED** [review]
This looks good to me and I trust from the other reviews that this is working and ready to merge! Thanks and good job everyone! <3
